vms/platformvm/txs/fee/calculator.go

```diff
@@ -0,0 +1,113 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package fee
+
+import (
+	"time"
+
+	"github.com/ava-labs/avalanchego/utils/constants"
+	"github.com/ava-labs/avalanchego/vms/platformvm/txs"
+	"github.com/ava-labs/avalanchego/vms/platformvm/upgrade"
+)
+
+var _ txs.Visitor = (*Calculator)(nil)
+
+type Calculator struct {
+	// Pre E-fork inputs
+	upgrades  upgrade.Times
+	staticCfg StaticConfig
+	chainTime time.Time
+
+	// outputs of visitor execution
+	Fee uint64
+}
+
+func NewStaticCalculator(cfg StaticConfig, ut upgrade.Times, chainTime time.Time) *Calculator {
+	return &Calculator{
+		upgrades:  ut,
+		staticCfg: cfg,
+		chainTime: chainTime,
+	}
+}
+
+func (fc *Calculator) AddValidatorTx(*txs.AddValidatorTx) error {
+	fc.Fee = fc.staticCfg.AddPrimaryNetworkValidatorFee
+	return nil
+}
+
+func (fc *Calculator) AddSubnetValidatorTx(*txs.AddSubnetValidatorTx) error {
+	fc.Fee = fc.staticCfg.AddSubnetValidatorFee
+	return nil
+}
+
+func (fc *Calculator) AddDelegatorTx(*txs.AddDelegatorTx) error {
+	fc.Fee = fc.staticCfg.AddPrimaryNetworkDelegatorFee
+	return nil
+}
+
+func (fc *Calculator) CreateChainTx(*txs.CreateChainTx) error {
+	fc.Fee = fc.staticCfg.GetCreateBlockchainTxFee(fc.upgrades, fc.chainTime)
+	return nil
+}
+
+func (fc *Calculator) CreateSubnetTx(*txs.CreateSubnetTx) error {
+	fc.Fee = fc.staticCfg.GetCreateSubnetTxFee(fc.upgrades, fc.chainTime)
+	return nil
+}
+
+func (*Calculator) AdvanceTimeTx(*txs.AdvanceTimeTx) error {
+	return nil // no fees
+}
+
+func (*Calculator) RewardValidatorTx(*txs.RewardValidatorTx) error {
+	return nil // no fees
+}
```

Comment:
We need to set `Fee` to zero or else we will still bill this as the previous tx's fee (assuming we're reusing a calculator, and not creating a new one per tx execution). Either way, I would recommend we be safe and just set the field to zero to avoid introducing a subtle bug even if this might be technically correct + add a regression test to cover this.

Reply:
I planned to re-instantiate the fee calculator at each use. This is not necessary pre E upgrade, but will be needed post E upgrade (with the dynamic calculator) because a calculator member will be in charge of cumulating the complexity across all txs in a block.
Still your suggestion helps safety, so done!